packages/arb-token-bridge-ui/src/hooks/useCompleteMultiChainTransactions.ts

```diff
@@ -0,0 +1,339 @@
+
+      if (_page === 0) {
+        return true
+      }
+
+      const txCount = latestTransactions.current[index]?.length ?? 0
+      // fetch next page for a chain pair if all fetched pages are full
+      // we check for >= in case some txs were included by initiating a transfer
+      return txCount / _page >= PAGE_SIZE
+    },
+    [latestTransactions, latestPage]
+  )
+
+  const { data, error } = useSWRImmutable(
+    address ? [direction, address, page] : null,
+    ([_direction, _address, _page]) => {
+      return Promise.all(
+        multiChainFetchList.map((c, index) => {
+          if (!shouldFetchPage(index)) {
+            return []
+          }
+
+          const params = {
+            sender: _address,
+            receiver: _address,
+            l1Provider: getProvider(c.parentChain),
+            l2Provider: getProvider(c.chain),
+            pageNumber: _page,
+            pageSize: PAGE_SIZE
+          }
+
+          return _direction === 'deposits'
+            ? fetchDepositList(params)
+            : fetchWithdrawalList(params)
+        })
+      )
+    }
+  )
+
+  useEffect(() => {
+    if (data) {
+      // if there is not a single full page in any of the data fetched, then there is no more to fetch
+      const shouldFetchNextPage = data.some((_, index) =>
+        shouldFetchPage(index)
+      )
+
+      // include the new data with the previously fetched data
+      // the data is grouped by chain pairs
+      setTransactions(prevTransactions => {
+        return data.map((transactionsForChainPair, chainPairIndex) => {
+          const prevTransactionsForChainPair = prevTransactions[chainPairIndex]
+          return [
+            ...(prevTransactionsForChainPair
+              ? prevTransactionsForChainPair
+              : []),
+            ...(transactionsForChainPair ? transactionsForChainPair : [])
+          ]
+        })
+      })
+
+      if (!shouldFetchNextPage) {
+        setLoading(false)
+        return
+      }
+
+      setPage(prevPage => prevPage + 1)
+    }
+  }, [data, shouldFetchPage])
+
+  return {
+    data: transactions.flat(),
+    loading,
+    error
+  }
+}
+
+/**
+ * Fetches transaction history only, without mapping additional info that would require a lot of RPC calls.
+ * The data is collected for multiple chain pairs, and sorted by date.
+ */
+const useMultiChainTransactionList = () => {
+  const deposits = useTransactionListByDirection('deposits')
+  const withdrawals = useTransactionListByDirection('withdrawals')
+
+  const groupedTransactions = [deposits, withdrawals]
+
+  // merge grouped transactions and sort them by date
+  const data = groupedTransactions
+    .flatMap(grp => grp.data)
+    .sort(sortByTimestampDescending)
+  // checks if any source is loading
+  const loading = groupedTransactions.map(grp => grp.loading).some(Boolean)
+  // get first error
+  const error =
+    groupedTransactions.map(grp => grp.error).filter(Boolean)[0] ?? null
```

Comment:
is `null` really needed? does it matter if it returns `undefined` instead?
`error` is of type `any` anyway